homeworks/hw1/hw1_template.py

```diff
@@ -1,58 +1,158 @@
+    with torch.no_grad():
+        predictions = model.predict(X).cpu().numpy()
+        y = y.cpu().numpy()
+
+        # Вычисляем матрицу ошибок
+        accuracy = accuracy_score(y_true=y, y_pred=predictions)
+        conf_matrix = confusion_matrix(y_true=y, y_pred=predictions)
+
     return predictions, accuracy, conf_matrix
 
 
-def train(model, criterion, optimizer, X_train, y_train, X_val, y_val, epochs):
-    ### YOUR CODE HERE: train the model and validate it every epoch on X_val, y_val
-    return model
+def train(model, criterion, optimizer, X_train, y_train, epochs, batch_size, X_val=None, y_val=None):
+    """Метод для обучения модели и проверки его на валидационной выборке"""
+
+    train_losses = []
+    val_losses = []
+
+    for epoch in range(epochs):
+        model.train()
+        train_loss = 0
+        for i in range(0, X_train.size(0), batch_size):
+            X_batch = X_train[i : i + batch_size]
+            y_batch = y_train[i : i + batch_size]
+
+            # Обновляем градиенты, forward, считаем loss и обновляем параметры
+            optimizer.zero_grad()
+            outputs = model(X_batch).squeeze()
+            loss = criterion(outputs, y_batch)
+            loss.backward()
+            optimizer.step()
+            train_loss += loss.item()
+
+        train_loss /= X_train.size(0) // batch_size
+        train_losses.append(train_loss)
+
+        if X_val is not None and y_val is not None:
+            # Оценка на валидационной выборке
+            model.eval()
+            val_loss = 0
+            with torch.no_grad():
+                for i in range(0, X_val.size(0), batch_size):
+                    X_batch_val = X_val[i : i + batch_size]
+                    y_batch_val = y_val[i : i + batch_size]
+
+                    outputs_val = model(X_batch_val).squeeze()
+                    loss_val = criterion(outputs_val, y_batch_val)
+                    val_loss += loss_val.item()
+
+            val_loss /= X_val.size(0) // batch_size  # Средний loss на валидационной выборке
+            val_losses.append(val_loss)
+
+            print(
+                f"Epoch {epoch}: Train Loss: {round(train_loss, 4)}"
+                f"| Epoch {epoch}: Validation Loss: {round(val_loss, 4)}"
+            )
+
+    return train_losses, val_losses
 
 
 def main(args):
-    ### YOUR CODE HERE
+    """Пайплайн для загрузки данных, обучения модели и ее инференса"""
 
     # Load data
+    X_train, y_train, X_val, y_val, X_test = load_data(
+        train_csv_path=args.train_csv, val_csv_path=args.val_csv, test_csv_path=args.test_csv
+    )
+
+    X_train, y_train = convert_to_tensor(X_train, y_train)
+    X_val, y_val = convert_to_tensor(X_val, y_val)
+    X_test = convert_to_tensor(X_test)
 
     # Initialize model
+    model, criterion, optimizer = init_model(model_params, args.lr)
 
     # Train model
+    _, _ = train(model, criterion, optimizer, X_train, y_train, args.num_epoches, args.batch_size, X_val, y_val)
```

Comment:
Minor: no need to write `_, _ = train(..)` if you don't use function call results, just call a function `train(..)`

Reply:
Thanks! I didn't think about it.